zebra-chain/src/transaction.rs

```diff
@@ -778,6 +830,10 @@ impl Transaction {
             | Transaction::V5 {
                 sapling_shielded_data: None,
                 ..
+            }
+            | Transaction::V6 {
+                sapling_shielded_data: None,
+                ..
```

Comment:
Let's simplify the entire function:
```rust
   /// Access the deduplicated [`sapling::tree::Root`]s in this transaction,
    /// regardless of version.
    pub fn sapling_anchors(&self) -> Box<dyn Iterator<Item = sapling::tree::Root> + '_> {
        // This function returns a boxed iterator because the different
        // transaction variants end up having different iterator types
        match self {
            Transaction::V4 {
                sapling_shielded_data: Some(sapling_shielded_data),
                ..
            } => Box::new(sapling_shielded_data.anchors()),

            Transaction::V5 {
                sapling_shielded_data: Some(sapling_shielded_data),
                ..
            } => Box::new(sapling_shielded_data.anchors()),

            Transaction::V6 {
                sapling_shielded_data: Some(sapling_shielded_data),
                ..
            } => Box::new(sapling_shielded_data.anchors()),

        // No Spends (V1, V2, V3 and V4, V5, V6 without sapling shielded data)
        _ => Box::new(iter::empty()),
    }
}
```
(x4 for all sapling functions)

Reply:
@arya2 what do you think about this simplification across the entire file?
This will significantly reduce boilerplate without affecting semantics.

Reply:
Such simplification would work here and in similar places. However, if we add a feature flag for the V6 feature as we discussed:

```rust
V5 { ... } |
#[cfg(feature = "tx-v6")]
V6 { ... } => { ... }
```

It wouldn't compile. Here's a [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a667922d54497227cbaa0ad0eec0c962) to illustrate the issue.

I introduced the `"tx-v6"` feature flag in the next PR (#17, "Step 4"). However, it would definitely be better to include it in this PR ("Step 3") since I introduced V6 here (the absence of the `"tx-v6"` feature flag in this PR makes the review process confusing).

Anyway, we can't afford to use the `|` operator for such simplification during the transition period when the `"tx-v6"` feature flag is needed. Although I'm not completely sure if Rust doesn't have workarounds to achieve that.

And it was one of the reasons why I added the `tx_v5_and_v6` macro [here](https://github.com/QED-it/zebra/blob/zsa-integration-generics/zebra-chain/src/transaction.rs#L109) – to avoid such code duplication. However, the macro itself can be confusing because its usage doesn't follow the existing style for processing transaction version matches. I might need to remove it, and we can discuss this further.